sycl/include/sycl/queue.hpp

```diff
@@ -81,6 +81,27 @@ class queue_impl;
 inline event submitAssertCapture(queue &, event &, queue *,
                                  const detail::code_location &);
 #endif
+
+// Function to postprocess submitted command
+// Arguments:
+// bool IsKernel - true if the submitted command was kernel, false otherwise
+// bool KernelUsesAssert - true if submitted kernel uses assert, only
+//                         meaningful when IsKernel is true
+// event &Event - event after which post processing should be executed
+using SubmitPostProcessF = std::function<void(bool, bool, event &)>;
+
+struct SubmissionInfoImpl;
+
+class __SYCL_EXPORT SubmissionInfo {
+public:
+  SubmissionInfo() = default;
+
+  void SetPostProcessing(const SubmitPostProcessF &PostProcessorFunc);
+  void
+  SetSecondaryQueue(const std::shared_ptr<detail::queue_impl> &SecondaryQueue);
+
+  std::shared_ptr<SubmissionInfoImpl> impl = nullptr;
```

Comment:
Why "shared"? I also think I'd prefer something like
```
struct SumissionInfo {
    std::optional<...> &SecondaryQueue();
    std::optional<...> &PostProcessorFunc();

    SubmissionInfo(); // in .cpp, ensures impl is always non-nullptr.

private:
    std::unique_ptr<SubmissionInfoImpl> impl;
}
```

Alternatively, we can split getters into `const`/non-`const` and do some kind of copy-on-non-const-get init of the `unique_ptr`.

Reply:
I think we had a similar discussion about `unique_ptr` vs. `shared_ptr` before. From what I remember of it, `unique_ptr` could not easily be used for pimpl, as it would be unable to call the destructor of an incomplete type.

As for the reference returning function solution, I partially like that, but I think there's one minus to it: Currently we only instantiate the `impl` only when it is needed. If we changed it to these kinds of functions, we could still do that, but then when the runtime sources would get them, they would initialize `impl` without actually needing to. The source code could, of course, still check this, but that assumes that future code also does that, which would be easy to miss.

Reply:
Discussed this offline. I will investigate the possible use of `unique_ptr` here. As for the reference functions, I think we agree that we could do something like it, but I vaguely remember us not being able to use `std::optional` across the library boundary, due to some ABI differences for `std::optional` implementations. I will investigate alternatives.